design/open/kv-query-tool.md

```diff
@@ -0,0 +1,45 @@
+# KV Store Querying Tool
+
+## Scope
+
+Our newly introduced KV Store stores the data in the form of [`Partition Key`, `Key`, `Value`] triplets. The `partition key` and `key` fields are strings stored as byte arrays and the `value` is an object (various types) serialized as protobuf and stored as byte array.
+This data is not easily interpretable for human viewer, whereas our former SQL DB stored the data (mostly) as human readable data, making it easier for developers to look into the DB, and debug various situations. It also provided the same ease to insert data or alter existing data in the DB for experimentation and, in some cases, as a fix
+With our new KV Store this became less trivial and this proposal comes to offer a solution for that
+
+## Introducing `lakefs kv`
+
+The `lakefs` server contains all the logic needed to correctly read and update the KV Store. It generates objects into byte arrays with their protobuf encoding, on one hand, and rebuilds the objects from byte arrays with protobuf on the other, and, it holds all the logic to generate the relevant keys (e.g. `partition key` from a repository object, branch `key` from the branch object etc.)
+We will elevate this logic and into a new command that will provide the ability to read data items, scan keys and update values, using a new `lakefs` command - `kv`
+```
+lakefs kv <cmd> <args>
+```
+
+### Get/Scan/Set
+
+* `lakefs kv get <partition> <key>` </br>
+  Given a `partition key` and a `key` (both as strings), returns the corresponding object (json format)
+
+* `lakefs kv scan <partition> [<key>]` </br>
+  Given a `partition key` and an optional `key`, returns all object in that partition that keys are equal or greater to the specified `key` (no `key` will return all objects in the partition). Results will be formatted as a json array
+
+* `lakefs kv set <partition> <key> <value>` </br>
+  Given a `partition key`, a `key` and a json formatted object, sets the value for the given `key` in the given `partition`, as an encoded protobuf for the given object
+
+### Delete (?)
+* `lakefs kv delete <partition> <key>` </br>
+  Given a `partition key` and a `key`, deletes the corresponding entry from the KV Store
+  **Question:** Do we really want this?
+
+### The `<key>` is the Key
```

Comment:
We do not expose paths to the user and in some cases it will be impossible to determine it without internal knowledge.
For example: how does the user infer the path of a branch under a repository which uses the repository UUID?

Reply:
Yap, repository UUID is a killer. Maybe we can try to abstract it away from the user, i.e. perform another call to get the repository UUID when given the repository path?

Reply:
To save two calls you can use:
```repo scan <repo-name>``` -  without the key, it will translate it to the result and returns a table with the keys and all repository entities like when you're querying directly from the DB.

Reply:
1. AFAIU, this tool is not meant to be exposed to users. I cannot prevent the agile user from looking at the code, understanding the logic and use it, but this tool is meant to help **us** understanding and debugging a system. I can see a scenario where we ask/suggest a user to run several operations, in order to help us get more info of his system state, but I believe it will have to be lead by one of us, so the knowledge is here
2. Adding ad-hoc commands will require to introduce specific knowledge into this logic (i.e. `repo-scan` will require to understand this is a repository object and parse it as such. While this is definitely an option, it is not easily scalable (e.g. introducing new object we would like to scan, like StagingToken). I think that as a first step we should consider limiting oursleves to a KV-like interface

Reply:
Since this is internal tool, we can expect the user (us) to have the required knowledge to operate it correctly, e.g, get the repository best on the well-known `graveler` partition and either `scan` or the repopsitoryID, and from there, building the repository's partition based on the InstanceUID (that we will need to expose)
Alternatively, we can implement a `get repo-partition`, or `repo-scan` to easily overcome this, although this breaks the generality I had in mind in the first place. I guess it's a question we need to ask